File: workspace_python/ros2_ws/src/python_workspace/python_workspace/picture_node.py
import time, os
import cv2
import random
import rclpy
from rclpy.time import Time
from rclpy.node import Node
from rclpy.executors import MultiThreadedExecutor
from sensor_msgs.msg import Image
from std_msgs.msg import Header, String
from cv_bridge import CvBridge

from custom_interfaces.msg import ImageInput                            # CHANGE

# node to publish static image data to the /image topic. 
# used for testing the inference pipelines.
class PictureNode(Node):
    def __init__(self):
        super().__init__('picture_node')

        self.declare_parameter('static_image_path', '/home/usr/Desktop/ROS/assets/IMG_1822_14.JPG')
        self.declare_parameter('loop', 0)  # 0 = don't loop, >0 = # of loops, -1 = loop forever
        self.declare_parameter('frame_rate', 1)  # Desired frame rate for publishing

        
        self.static_image_path = self.get_parameter('static_image_path').get_parameter_value().string_value
        # Resolve the path programmatically
        if not os.path.isabs(self.static_image_path):
            self.static_image_path = os.path.join(os.getcwd(), self.static_image_path)

        self.loop = self.get_parameter('loop').get_parameter_value().integer_value
        self.frame_rate = self.get_parameter('frame_rate').get_parameter_value().integer_value

        self.bridge = CvBridge()

        self.image_list = self.get_images()
        self.loop_count = 0
        self.image_counter = 0

        self.input_image_publisher = self.create_publisher(ImageInput, 'input_image', 10)
        timer_period = 1/self.frame_rate  # publish every 0.5 seconds
        self.timer = self.create_timer(timer_period, self.publish_static_image)

           
    def get_images(self):
        """
        Returns a list of images in the form of ROS2 image messages from the path specified by the static_image_path parameter.
        """
        # You can technically read the img file as binary data rather than using cv2 to read it and then convert it into a ROS2 image message... and then converting it back to a numpy array in the inference node. But since the picture node is only for testing the MVP and we're not using the GPU yet the overhead here should not matter.
        if not os.path.exists(self.static_image_path):
            self.get_logger().error(f"Static image not found at {self.static_image_path}")
            raise FileNotFoundError(f"Static image not found at {self.static_image_path}")
        
        image_paths = []
        if os.path.isfile(self.static_image_path) \
            and (filename.endswith('.JPG') or filename.endswith('.png')):
            filename = os.path.join(self.static_image_path, filename)
            image_paths.append(filename)
        elif os.path.isdir(self.static_image_path):
            for filename in os.listdir(self.static_image_path):
                if filename.endswith('.JPG') or filename.endswith('.png'):
                    filename = os.path.join(self.static_image_path, filename)
                    image_paths.append(filename)

        if len(image_paths) == 0:
            self.get_logger().error(f"No images found at {self.static_image_path}")
            return
        
        images = []
        counter = 0
        for filename in image_paths:

            image = cv2.imread(filename, cv2.IMREAD_COLOR)
            ros_image = self.bridge.cv2_to_imgmsg(image, encoding="bgr8")
            if image is None:
                self.get_logger().error(f"Failed to read image: {filename}")
                raise FileNotFoundError(f"Failed to read image: {filename}")
            
            image_msg = ImageInput()
            # image_msg.header.frame_id = f'camera_frame{counter}'
            image_msg.image = ros_image
            image_msg.velocity = random.uniform(0, 2)

            counter +=1

            images.append(image_msg)

        return images

    
    def publish_static_image(self):
        """
        Publishes static images to the /image topic
        """
        array_size = len(self.image_list)

        # publish the image on the current pos to the image topic
        if self.loop == -1 or self.loop_count < self.loop:
            position = self.image_counter % array_size
            self.get_logger().info(f"Publishing image {position + 1}/{array_size}")
            self.input_image_publisher.publish(self.image_list[position])

            self.image_counter += 1
            self.loop_count = self.image_counter // array_size
        else:
            # stop the timer/quit the node
            self.timer.cancel()
            self.get_logger().info("Finished publishing images")
            self.destroy_node()

def main(args=None):
    rclpy.init(args=args)
    picture_node = PictureNode()
    executor = MultiThreadedExecutor(num_threads=1)
    executor.add_node(picture_node)
    
    try:
        executor.spin()
    finally:
        executor.shutdown()
        picture_node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: ros2_ws/src/cpp_package/src/picture_node.cpp
#include <chrono>
#include <fstream>
#include <iostream>
#include <memory>
#include <random>
#include <string>
#include <vector>
#include <filesystem>
#include <opencv2/opencv.hpp>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "std_msgs/msg/header.hpp"
#include "custom_interfaces/msg/image_input.hpp"
#include "cv_bridge/cv_bridge.h"

using namespace std::chrono_literals;
namespace fs = std::filesystem;


// Parameter Handling:
// Parameters such as static_image_path, loop, and frame_rate are declared and fetched using the ROS2 parameter system.

// Image Preprocessing:
// Example preprocessing includes resizing the image. Additional transformations can be added as needed.

// ROS2 Message Conversion:
// OpenCV images are converted to ROS2 messages using the cv_bridge library.

// Dynamic Publishing:
// Publishes images at a configurable frame rate using a timer.

// Error Handling:
// Ensures that the node throws meaningful errors if the image path is invalid or no images are found.

// Random Velocity:
// A random velocity is generated for each image for testing purposes.

class PictureNode : public rclcpp::Node
{
public:
    PictureNode() : Node("picture_node"), loop_count_(0), image_counter_(0)
    {
        // Declare parameters
        declare_parameter<std::string>("static_image_path", "/home/user/ROS/assets/maize/IMG_1822_14.JPG");
        declare_parameter<int>("loop", -1);  // 0 = don't loop, >0 = # of loops, -1 = loop forever
        declare_parameter<int>("frame_rate", 1); // Desired frame rate for publishing

        static_image_path_ = get_parameter("static_image_path").as_string();
        loop_ = get_parameter("loop").as_int();
        frame_rate_ = get_parameter("frame_rate").as_int();

        if (!fs::exists(static_image_path_))
        {
            RCLCPP_ERROR(this->get_logger(), "Static image path does not exist: %s", static_image_path_.c_str());
            throw std::runtime_error("Static image path not found");
        }

        // Load images
        load_images();

        // Publisher
        image_publisher_ = this->create_publisher<custom_interfaces::msg::ImageInput>("input_image", 10);

        // Timer
        auto timer_period = std::chrono::duration<double>(1.0 / frame_rate_);
        timer_ = this->create_wall_timer(timer_period, std::bind(&PictureNode::publish_static_image, this));
    }

private:
    void load_images()
    {
        if (fs::is_regular_file(static_image_path_))
        {
            if (is_valid_image(static_image_path_))
            {
                images_.push_back(cv::imread(static_image_path_));
            }
        }
        else if (fs::is_directory(static_image_path_))
        {
            for (const auto &entry : fs::directory_iterator(static_image_path_))
            {
                if (entry.is_regular_file() && is_valid_image(entry.path()))
                {
                    images_.push_back(cv::imread(entry.path()));
                }
            }
        }

        if (images_.empty())
        {
            RCLCPP_ERROR(this->get_logger(), "No valid images found in path: %s", static_image_path_.c_str());
            throw std::runtime_error("No valid images found");
        }
    }

    bool is_valid_image(const fs::path &path)
    {
        auto ext = path.extension().string();
        return ext == ".JPG" || ext == ".jpg" || ext == ".png";
    }

    void publish_static_image()
    {
        if (loop_ == -1 || loop_count_ < loop_)
        {
            size_t position = image_counter_ % images_.size();
            RCLCPP_INFO(this->get_logger(), "Publishing image %zu/%zu", position + 1, images_.size());

            // Prepare the image
            auto raw_image = images_[position];
            auto preprocessed_image = preprocess_image(raw_image);

            // Convert to ROS2 messages
            auto raw_msg = cv_bridge::CvImage(std_msgs::msg::Header(), "bgr8", raw_image).toImageMsg();
            auto preprocessed_msg = cv_bridge::CvImage(std_msgs::msg::Header(), "bgr8", preprocessed_image).toImageMsg();

            // Publish message
            custom_interfaces::msg::ImageInput image_input_msg;
            image_input_msg.header.stamp = this->get_clock()->now();
            image_input_msg.header.frame_id = "static_image";
            image_input_msg.raw_image = *raw_msg;
            image_input_msg.preprocessed_image = *preprocessed_msg;
            image_input_msg.velocity = generate_random_velocity();

            image_publisher_->publish(image_input_msg);

            // Update counters
            image_counter_++;
            loop_count_ = image_counter_ / images_.size();
        }
        else
        {
            // Stop publishing
            timer_->cancel();
            RCLCPP_INFO(this->get_logger(), "Finished publishing images");
            rclcpp::shutdown();
        }
    }

    cv::Mat preprocess_image(const cv::Mat &image)
    {
        cv::Mat processed;
        cv::resize(image, processed, cv::Size(224, 224)); // Example preprocessing
        return processed;
    }

    float generate_random_velocity()
    {
        static std::random_device rd;
        static std::mt19937 gen(rd());
        static std::uniform_real_distribution<> dis(0.0, 1.0);
        return dis(gen);
    }

    // Parameters
    std::string static_image_path_;
    int loop_;
    int frame_rate_;

    // Images and counters
    std::vector<cv::Mat> images_;
    size_t loop_count_;
    size_t image_counter_;

    // ROS2 publishers and timers
    rclcpp::Publisher<custom_interfaces::msg::ImageInput>::SharedPtr image_publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto picture_node = std::make_shared<PictureNode>();
    rclcpp::spin(picture_node);
    rclcpp::shutdown();
    return 0;
}
